File: 프로그래머스/lv1/118666. 성격 유형 검사하기/성격 유형 검사하기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(vector<string> survey, vector<int> choices) {
    string answer = "";
    pair<pair<char, char>, int> type[4] = {{{'R', 'T'}, 0}, {{'C', 'F'}, 0},
                                           {{'J', 'M'}, 0}, {{'A', 'N'}, 0}};
    int score[8] = {0,3,2,1,0,-1,-2,-3};
    for (int i = 0; i < survey.size(); i++) {
        int flag = 1;
        if (survey[i][0] > survey[i][1])
            flag = -1;
        if (survey[i] == "RT" || survey[i] == "TR")
            type[0].second += score[choices[i]]*flag;
        else if (survey[i] == "CF" || survey[i] == "FC")
            type[1].second += score[choices[i]]*flag;
        else if (survey[i] == "JM" || survey[i] == "MJ")
            type[2].second += score[choices[i]]*flag;
        else
            type[3].second += score[choices[i]]*flag;
    }

    for (int i = 0; i < 4; i++) {
        if (type[i].second >= 0)
            answer += type[i].first.first;
        else
            answer += type[i].first.second;
    }
    return answer;
}

File: 프로그래머스/lv1/118666. 성격 유형 검사하기/성격 유형 검사하기.swift
import Foundation

func solution(_ survey:[String], _ choices:[Int]) -> String {
    var answer = String()
    let type = ["RT", "CF", "JM", "AN"]
    var check: [Character: Int] = [:]
    
    type.forEach {
        check[$0.first!] = 0
        check[$0.last!] = 0
    }
    zip(survey, choices).forEach {
        if $0.1 < 4 { check[$0.0.first!]! += (4-$0.1) }
        else { check[$0.0.last!]! += ($0.1-4) }
    }
    
    for i in 0..<4  {
        answer += check[type[i].first!]! >= check[type[i].last!]! ? String(type[i].first!) : String(type[i].last!)
    }
     
    return answer
}